opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/lz4_chunked_compression.rs

```diff
@@ -0,0 +1,149 @@
+use lz4_flex::block::{compress_into, get_maximum_output_size};
+
+/// Compresses input data in 64 KiB chunks using LZ4, writing each chunk's compressed data to a single
+/// pre-allocated buffer. Each chunk in the output is prefixed by a 4-byte (little-endian) length header
+/// indicating the size of the compressed data that follows. This chunked format allows easy and efficient
+/// decompression, as each compressed chunk can be read independently by first reading its 4-byte length,
+/// then the compressed payload.
+///
+/// # Output Buffer Layout (Block Diagram)
+/// ```text
+///  |<--Chunk 1-->|<--Chunk 2-->| ... |<--Chunk N-->|
+///  +----+--------+----+-------+----+-----+
+///  | 04 |....    | 03 |...    | 02 |..   |
+///  +----+--------+----+-------+----+-----+
+///
+///  Where:
+///    04, 03, 02  = 4-byte little-endian u32 chunk length headers
+///    ...., ..., .. = compressed LZ4 data for each chunk (length matches header)
+///    N     = number of chunks = (input.len() / CHUNK_SIZE) + (if input.len() % CHUNK_SIZE > 0 { 1 } else { 0 })
+///
+/// Example for 3 chunks (lengths 4, 3, 2 bytes):
+///  +----+--------+----+-------+----+-----+
+///  | 04 |....    | 03 |...    | 02 |..   |
+///  +----+--------+----+-------+----+-----+
+///   ^len1 data1   ^len2 data2   ^len3 data3
+/// ```
+///
+/// # Notes
+/// - This chunked format is **not required by LZ4** itself, but is a common convention to allow boundary detection.
+/// - Decompression requires reading 4 bytes, then decompressing the next `len` bytes, then repeating.
+/// - This function is allocation-efficient: only two buffers are allocated (one for output, one for temp).
+/// # TODO
+/// - Investigate if true in-place compression is possible when the input is no longer needed.
+///   This might let us use the input buffer as the temporary compression buffer, potentially reducing heap allocations further,
+///   but care must be taken: LZ4 does not natively support in-place compression, and the compressed size may be larger than the input.
+///   If single-buffer "in-place" compression is possible (e.g., with unsafe or buffer aliasing), document or implement it here.
+#[allow(dead_code)]
+pub(crate) fn lz4_chunked_compression(input: &[u8]) -> Vec<u8> {
+    const CHUNK_SIZE: usize = 64 * 1024;
+
+    // Pre-allocate an output buffer large enough for the worst-case total output size:
+    // Each chunk may require up to get_maximum_output_size(CHUNK_SIZE) bytes for compressed data,
+    // plus 4 bytes for the length header per chunk.
+    let mut output = Vec::with_capacity(
```

Comment:
Also, we should set a limit for the maximum size that we would allocate a vec with. It shouldn't be unbounded.

Reply:
If the input size is bounded by upstream, we can avoid adding maximum size for output here?